### PR TITLE
ref: Remove data field from event model

### PR DESCRIPTION
**Pull request by lynnagara**
This PR aims to separate writing events to Nodestore from writing events to Postgres.

Previously nodestore data was automatically written when an event is saved. With these changes calling event.save() will no longer save event bodies to nodestore, instead nodedata.save() needs to be explicitly called in order to save nodestore data.

Similarly to before, accessing data on an event object will force a nodestore load if data is not yet loaded.

We are no longer populating the data field in the event table since events are always looked up by a hash of the project ID and event UUID.

Note that we are not writing the "_ref" and "_ref_version" fields to nodestore any more.

This is the first step in preparation for completely discontinuing writing to the Event table.

### CODE REVIEW

File: src/sentry/south_migrations/0506_auto__del_field_event_data.py [hunk 1]
```diff
@@ -0,0 +1,1467 @@
...
+
+class Migration(SchemaMigration):
+
+    # Flag to indicate if this migration is too risky
+    # to run online and needs to be coordinated for offline
+    is_dangerous = True
+
+    def forwards(self, orm):
+        # Deleting field 'Event.data'
+        db.delete_column('sentry_message', 'data')
```
Comment (markstory) [inline src/sentry/south_migrations/0506_auto__del_field_event_data.py]: Won't deleting the column happen before the code is rolled out causing problems?

File: src/sentry/models/event.py [hunk 2]
```diff
@@ -633,6 +619,17 @@ class Meta:
 
     __repr__ = sane_repr("project_id", "group_id")
 
+    def __init__(self, *args, **kwargs):
+        node_data = kwargs.pop("node_data", None)
```
Comment (fpacifici) [inline src/sentry/models/event.py]: I think having both node_data and data is a bit confusing, thus it is hard to assess what the behavior is.
What's the difference between data and node_data ? When would we use one or the other ?

File: src/sentry/tasks/post_process.py [hunk 3]
```diff
@@ -139,7 +140,7 @@ def post_process_group(event, is_new, is_regression, is_sample, is_new_group_env
 
         # Re-bind node data to avoid renormalization. We only want to
         # renormalize when loading old data from the database.
-        event.data = EventDict(event.data, skip_renormalization=True)
+        # event.data = EventDict(event.data, skip_renormalization=True)
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply src/sentry/tasks/post_process.py]: I believe this was caused by the way NodeField was loading data. I'm not 100% confident about this though. @untitaker Does this look reasonable to you?

File: src/sentry/models/event.py (hunk 2 above)
Comment (lynnagara) [reply src/sentry/models/event.py]: Event accepts either data or NodeData now, I updated this to check the type instead

Comment (fpacifici) [reply src/sentry/models/event.py]: I think my questions is more about the which use cases should pass data and which ones should pass node_data.

File: src/sentry/event_manager.py [hunk 4]
```diff
@@ -431,13 +432,13 @@ def _get_event_instance(self, project_id=None):
         date = datetime.fromtimestamp(recorded_timestamp)
         date = date.replace(tzinfo=timezone.utc)
         time_spent = data.get("time_spent")
-
-        data["node_id"] = Event.generate_node_id(project_id, event_id)
-
+        node_data = NodeData(
+            None, None, data=EventDict(data, skip_renormalization=True), wrapper=EventDict
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: Isn't this calling the EventDict wrapper passing an object that is already an EventDict ?

File: src/sentry/models/event.py [hunk 5]
```diff
@@ -633,6 +619,16 @@ class Meta:
 
     __repr__ = sane_repr("project_id", "group_id")
 
+    def __init__(self, *args, **kwargs):
```
Comment (fpacifici) [inline src/sentry/models/event.py]: I think this allowing this to take the same argument in two different types is going to be very confusing.
I would add factory methods, one for NodeData and ther other for the other type you allow, and please add a docstring to each explaining when to use them. 

But taking a step back, can we converge into one type only ?

File: src/sentry/event_manager.py [hunk 6]
```diff
@@ -559,8 +560,8 @@ def save(self, project_id, raw=False, assume_normalized=False):
         # We need to swap out the data with the one internal to the newly
         # created event object
         event = self._get_event_instance(project_id=project_id)
-        self._data = data = event.data.data
-
+        # self._data = data = event.data.data
+        assert self._data == data == event.data.data
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: Overall I am not sure I get why we need to change the type of Event.data from EventDict to NodeData.
Also what's the reason for these three representations of event to coexist ? Are we modifying them independently in such a way that we still have to preserve all of them ?

**Review by fpacifici — CHANGES_REQUESTED** [review]
Still a few clarifications are needed.
Also, this is changing some very delicate logic in envet_manager. How did you test this PR ?

File: src/sentry/models/event.py (hunk 2 above)
Comment (lynnagara) [reply src/sentry/models/event.py]: Right now only event manager passes node data. It's kind of a hack to avoid further refetching node data during post process, since it gets a NodeData with no node_id it won't fetch from nodestore anymore.

Comment (fpacifici) [reply src/sentry/models/event.py]: Why do we need to avoid refetching node data in post process?

File: src/sentry/event_manager.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Looking back at https://github.com/getsentry/sentry/commit/54764bf9d9ed8473784fbecbb86badeaaad10d97 I believe it's needed to renormalize node data like we do with SnubaEvent

File: src/sentry/event_manager.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: My goal is for accessing event.data to stay unchanged. Prior to this change NodeField took care of always returning you a NodeData instance https://github.com/getsentry/sentry/blob/master/src/sentry/db/models/fields/node.py#L219.

I'm not sure if we'll be able to do away with any of this in future, likely not anytime soon though since RawEvent still uses NodeData in the "old" way

Comment (untitaker) [reply src/sentry/event_manager.py]: The type of event.data was always `NodeData`. `EventDict` was wrapped by NodeData.

The only purpose of `EventDict` was to apply renormalization transparently when `NodeField` was loading the data from nodestore. I would love to get rid of this implicit hack if we know all the places where `event.data` is accessed. The assumption at the time was that we don't because of plugins.

Please be very careful when rolling this out. Applying renormalization too often can cause a slowdown of post-process-forwarder and create a backlog that impacts how fast events are stored and indexed. We have observed such a backlog, and concluded that pickling and unpickling (or something like it) of the event object causes renormalization to be applied again when it was not necessary. That's why sometimes we explicitly re-set event.data to itself with the `skip_renormalization` flag.

The other thing that can happen during deploy is that we do not renormalize when we should. That can cause UI crashes when old, partially broken events (that violate the schema, or an invariant in the data that the UI relies on). This is not a big problem because there are not many events like that, but it's hard to monitor for. More often the UI does not crash, but the API endpoint for the issue details page. That one is usually easy to monitor and easy to fix.

ccing @mitsuhiko on this because he has a better understanding of how NodeData works.

File: src/sentry/models/event.py [hunk 7]
```diff
@@ -613,13 +606,6 @@ class Event(EventCommon, Model):
     platform = models.CharField(max_length=64, null=True)
     datetime = models.DateTimeField(default=timezone.now, db_index=True)
     time_spent = BoundedIntegerField(null=True)
-    data = NodeField(
```
Comment (fpacifici) [inline src/sentry/models/event.py]: We are not going to delete nodes when we delete an event with this PR, right ?
This was happening here:
https://github.com/getsentry/sentry/blob/master/src/sentry/db/models/fields/node.py#L179-L184

If I did not get this wrong, we need to find an intermediate solution before the synchronized consumer solution is ready.

File: src/sentry/event_manager.py [hunk 8]
```diff
@@ -523,8 +522,6 @@ def save(self, project_id, raw=False, assume_normalized=False):
         # We need to swap out the data with the one internal to the newly
         # created event object
         event = self._get_event_instance(project_id=project_id)
-        self._data = data = event.data.data
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: I am quite concerned about this line.
My comment on the previous version (where this became an assertion) was about using only one of the three in the rest of the method, but there is still code that writes and reads from both data and event.data.data (see event.get_hashes). 
Do you know why this line existed in the first place ?
In the new implementation self._data and event.data.data are not the same. self._data is a CanonicalKeyDict while event.data.data is an EventDict (which is a subclass). Is this intended ?

File: src/sentry/event_manager.py [hunk 9]
```diff
@@ -745,7 +739,10 @@ def save(self, project_id, raw=False, assume_normalized=False):
         # save the event
         try:
             with transaction.atomic(using=router.db_for_write(Event)):
+                event.set_data(EventDict(data, skip_renormalization=True))
```
Comment (fpacifici) [inline src/sentry/event_manager.py]: Why is this needed ? The entire method worked with data which had been assigned from self._data, which is the same assigned to event.data at line 406.

**Review by fpacifici — COMMENTED** [review]
How confident are you that nothing (inside and outside this repo) depends on the content of sentry_message.data ? This PR will stop writing in that field, we need to be sure nothing is reading/writing there anymore.

I would consider reducing the risk of this PR by splitting the logic that stops automatically saving on nodestore (node.py line 235) from the logici that manages the field on the Event class. Essentially you can remove the automatic management of nodestore and leave the data field in postgres intact and still save on nodestore manually, which is actually what you want. What do you think about this ?

Also I think a review by @mitsuhiko  would be desirable since he has a lot more context on this code.

File: src/sentry/tasks/post_process.py [hunk 10]
```diff
@@ -137,9 +136,8 @@ def post_process_group(event, is_new, is_regression, is_new_group_environment, *
         from sentry.rules.processor import RuleProcessor
         from sentry.tasks.servicehooks import process_service_hook
 
-        # Re-bind node data to avoid renormalization. We only want to
-        # renormalize when loading old data from the database.
-        event.data = EventDict(event.data, skip_renormalization=True)
+        # Clear node id to prevent reloading data from nodestore in post process
+        event.clear_node_id()
```
Comment (lynnagara) [inline src/sentry/tasks/post_process.py]: This definitely needs be double checked by someone, but this is the part of re-binding node data that I believe is important since clearing the ID prevents any further nodestore access or attempts to rebind data.

File: src/sentry/models/event.py (hunk 5 above)
Comment (lynnagara) [reply src/sentry/models/event.py]: The `clear_node_id` method attempts to make this a bit clearer

File: src/sentry/models/event.py (hunk 7 above)
Comment (lynnagara) [reply src/sentry/models/event.py]: Good point. I think the easiest solution is to update deletions to use generate_node_id method instead of this field first. I'm worried that changing the event model to keep this as well as the new data field we are adding here might affect plugins and other things downstream that would get an event object with a slightly different shape than before.

File: src/sentry/event_manager.py (hunk 9 above)
Comment (lynnagara) [reply src/sentry/event_manager.py]: Fair point, this was needed since data diverged from event.data.data when we dropped line 406. Reverted these bits back.

**Comment by lynnagara** [comment]
On second thoughts it might be worth keeping ref and ref_version in nodestore (at least for now) for backwards compatibility

**Comment by lynnagara** [comment]
Closed in favour of https://github.com/getsentry/sentry/pull/15004
